core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -312,15 +314,20 @@ public TableMetadata updatePartitionSpec(PartitionSpec newPartitionSpec) {
     Preconditions.checkArgument(defaultSpecId != newDefaultSpecId,
         "Cannot set default partition spec to the current default");
 
+    // start from last partition spec's fieldId
+    AtomicInteger nextPartitionFieldId = new AtomicInteger(lastPartitionFieldId);
     ImmutableList.Builder<PartitionSpec> builder = ImmutableList.<PartitionSpec>builder()
         .addAll(specs);
     if (!specsById.containsKey(newDefaultSpecId)) {
       // get a fresh spec to ensure the spec ID is set to the new default
-      builder.add(freshSpec(newDefaultSpecId, schema, newPartitionSpec));
+      PartitionSpec freshSpec = freshSpecWithAssignIds(newDefaultSpecId, schema, schema, newPartitionSpec,
+          nextPartitionFieldId, partitionFieldIdByColumnName);
```

Comment:
I think this should work like `updateSchema`, where a different class is responsible for reassigning IDs. The `TableMetadata` class should validate consistency and help with tracking (like the snapshot log) but it shouldn't modify other objects that are passed in, like schemas, snapshots, and partition specs.

Reply:
@rdblue 
are we referring to this
`Schema freshSchema = TypeUtil.assignFreshIds(schema, lastColumnId::incrementAndGet);`

as `TypeUtil.assignFreshIds` assign id to the schema ?